src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,283 @@
+        Some(ident) => ident.span,
+        _ => tcx.def_span(def_id),
+    };
+
+    let mut err = tcx.sess.struct_span_err(fn_span, "item has unused generic parameters");
+
+    let mut unused_parameters = BitSet::new_filled(used_parameters.domain_size());
+    unused_parameters.subtract(used_parameters);
+    debug!("emit_used_generic_parameters_error: unused_parameters={:?}", unused_parameters);
+
+    for index in unused_parameters.iter() {
+        let param = ty::ParamTy::new(index, Symbol::intern(&format!("P{}", index)));
+        // Can safely call `type_param` as all non-type parameters are considered used.
+        let generic_param_def = generics.type_param(&param, tcx);
+        debug!(
+            "emit_used_generic_parameters_error: generic_param_def={:?} param={:?}",
+            generic_param_def, param
+        );
+        let def_span = tcx.def_span(generic_param_def.def_id);
+        err.span_label(
+            def_span,
+            &format!("generic parameter `{}` is unused", generic_param_def.name),
+        );
+    }
+
+    err.emit();
+}
+
+/// Visitor used to aggregate generic parameter uses.
+struct UsedGenericParametersVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    body: ReadOnlyBodyAndCache<'a, 'tcx>,
+    used_parameters: &'a mut BitSet<u32>,
+}
+
+impl<'a, 'tcx> UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_foldable<T: std::fmt::Debug + TypeFoldable<'tcx>>(&mut self, ty: T) {
+        debug!("visit_foldable: ty={:?}", ty);
+        // Exit early if this type doesn't use any type parameters.
+        if !ty.has_param_types() {
+            return;
+        }
+
+        // Use `TypeVisitor` to find any type parameters.
+        ty.visit_with(self);
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_local_decl(&mut self, local: Local, local_decl: &LocalDecl<'tcx>) {
+        debug!("visit_local_decl: local_decl={:?}", local_decl);
+        let ty = local_decl
+            .ty
+            .builtin_deref(false)
+            .map(|type_and_mut| type_and_mut.ty)
+            .unwrap_or(local_decl.ty);
+        debug!("visit_local_decl: ty={:?}", ty);
+        // Skip visiting the closure that is currently being processed. This only happens
+        // because the first argument to the closure is a reference to itself and that
+        // will call `visit_substs`, resulting in each generic parameter captured being
+        // considered used by default.
+        match ty.kind {
+            ty::Closure(closure_def_id, _) if closure_def_id == self.def_id => {
+                debug!("visit_local_decl: skipping closure substs");
+                return;
+            }
+            _ => {}
+        }
+
+        self.super_local_decl(local, local_decl);
+    }
+
+    fn visit_const(&mut self, c: &&'tcx ty::Const<'tcx>, _: Location) {
+        debug!("visit_const: c={:?}", c);
+        // Necessary to detect `T::ASSOC_CONST`.
```

Comment:
It's not that specific, this is just as important as `visit_ty`, so the comment is slightly misleading.

Reply:
I've fixed this in dd24b5b.